### PR TITLE
Stableswap: Simple parts of LPing

### PR DESCRIPTION
**Pull request by ValarDragon**
Component of: #1406 

## What is the purpose of the change

This PR abstracts the Balancer logic for exit pool, and JoinPoolNoSwap logic to internal code usable by both balancer and stableswap.

This PR then adds these functions to stableswap. What is missing from stableswap after this is then the logic for single asset LP'ing. I plan on doing that tomorrow, it will be more logically intense, so I was thinking of doing it in a separate PR to keep this one simple to review, if folks think that is a good plan. 

## Testing and Verifying

* Balancer changes are covered by existing tests.
* This lends some degree of confidence to their correctness within stableswap, as this code is reused.
* Stableswap needs similar tests. This should imo be done by abstracting the inverse relation tests to allow more pools as input. Cref #{TODO create an issue}

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? No
  - How is the feature or change documented? Needs documentation within stableswap spec

### CODE REVIEW

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+	for _, asset := range poolLiquidity {
+		// round down here, due to not wanting to over-exit
+		exitAmt := shareOutRatio.MulInt(asset.Amount).TruncateInt()
+		if exitAmt.LTE(sdk.ZeroInt()) {
+			continue
+		}
+		if exitAmt.GTE(asset.Amount) {
+			return sdk.Coins{}, errors.New("too many shares out")
+		}
```
Comment (ValarDragon) [inline x/gamm/pool-models/internal/cfmm_common/lp.go:43]: This is new, but solves an edge case bug with potential rounding errors in the share approximation multiplication.

**Comment by ValarDragon** [comment]
Next steps:

* Add tests for newly introduced code:
  * BinarySearch (a basic test, more extensive can be done in follow-up)
  * SingleAssetJoin logic
* SwapFee for SingleAssetJoin
* Unify LP logic between Balancer and Stableswap, using more internal methods
* Consider adding a minimum `remCoins` before doing a singelAssetJoin for multi-asset. (Informs decisions on #1443)

File: osmoutils/binary_search.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	if e.AdditiveTolerance.IsZero() {
+		if expected.Equal(actual) {
+			return 0
+		} else {
+			return comparisonSign
+		}
+	}
+
+	// Check additive tolerance equations
+	if !e.AdditiveTolerance.IsNil() && !e.AdditiveTolerance.IsZero() {
```
Comment (p0mvn) [inline osmoutils/binary_search.go:44]: Should we check `e.AdditiveTolerance.IsNil()` before `e.AdditiveTolerance.IsZero()` above?

Comment (p0mvn) [reply osmoutils/binary_search.go:44]: Also, `!e.AdditiveTolerance.IsZero()` is always true at this point so it can be removed

File: x/gamm/pool-models/stableswap/amm.go [hunk 3]
```diff
@@ -352,11 +356,51 @@ func (pa *Pool) calcInAmtGivenOut(tokenOut sdk.Coin, tokenInDenom string, swapFe
...
+
+	// Add all exact coins we can (no swap). ctx arg doesn't matter for Stableswap
+	numShares, remCoins, err := cfmm_common.MaximalExactRatioJoin(pa, sdk.Context{}, tokensIn)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.NewCoins(), err
+	}
+	pa.updatePoolForJoin(tokensIn.Sub(remCoins), numShares)
+
+	for _, coin := range remCoins {
+		// TODO: Perhaps add a method to skip if this is too small.
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/amm.go:397]: let's create an issue please

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+		// so now due to correctness of exitPool, we exitPool and swap all remaining assets to base asset
+		exitFee := sdk.ZeroDec()
+		exitedCoins, err := poolWithUpdatedLiquidity.ExitPool(ctx, sharesIn, exitFee)
+		if err != nil {
+			return sdk.Int{}, err
+		}
+
+		return swapAllCoinsToSingleAsset(poolWithUpdatedLiquidity, ctx, exitedCoins, swapToDenom)
+	}
+	// TODO: Come back and revisit err tolerance
```
Comment (p0mvn) [inline x/gamm/pool-models/internal/cfmm_common/lp.go:140]: let's create an issue please

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+//
+// This implementation all of pool.GetTotalPoolLiquidity, pool.ExitPool, and pool.SwapExactAmountIn to not
+// require any updates to state, and instead only do updates based upon the pool struct.
+func BinarySearchSingleAssetJoin(
+	pool types.PoolI,
+	tokenIn sdk.Coin,
+	poolWithAddedLiquidityAndShares func(newLiquidity sdk.Coin, newShares sdk.Int) types.PoolI,
+) (numLPShares sdk.Int, err error) {
+	// use dummy context
+	ctx := sdk.Context{}
```
Comment (p0mvn) [inline x/gamm/pool-models/internal/cfmm_common/lp.go:114]: Why don't we propagate the context from the caller?

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	// Need to get something that makes the result correct within 1 LP share
+	// If we fail to reach it within maxIterations, we return an error
+	correctnessThreshold := sdk.NewInt(2)
+	maxIterations := 512
```
Comment (p0mvn) [inline x/gamm/pool-models/internal/cfmm_common/lp.go:118]: Why these values in particular? Also, please create constants instead

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+// Thanks to CFMM path-independence, we can estimate slippage with these swaps to be sure to get the right numbers here.
+// (by path-independence, swap all of B -> A, and then swap all of C -> A will yield same amount of A, regardless
+// of order and interleaving)
+//
+// This implementation all of pool.GetTotalPoolLiquidity, pool.ExitPool, and pool.SwapExactAmountIn to not
+// require any updates to state, and instead only do updates based upon the pool struct.
+func BinarySearchSingleAssetJoin(
+	pool types.PoolI,
+	tokenIn sdk.Coin,
+	poolWithAddedLiquidityAndShares func(newLiquidity sdk.Coin, newShares sdk.Int) types.PoolI,
```
Comment (p0mvn) [inline x/gamm/pool-models/internal/cfmm_common/lp.go:111]: Why do we need to pass in this callback when `pool` that is used by the caller to create a copy is accessible within the scope of this function?

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	maxIterations := 512
+	// upperbound of number of LP shares = existingShares * tokenIn.Amount / pool.totalLiquidity.AmountOf(tokenIn.Denom)
+	existingTokenLiquidity := pool.GetTotalPoolLiquidity(ctx).AmountOf(tokenIn.Denom)
+	existingLPShares := pool.GetTotalShares()
+	LPShareUpperBound := existingLPShares.Mul(tokenIn.Amount).ToDec().QuoInt(existingTokenLiquidity).Ceil().TruncateInt()
+	LPShareLowerBound := sdk.ZeroInt()
+
+	// Creates a pool with tokenIn liquidity added, where it created `sharesIn` number of shares.
+	// Returns how many tokens you'd get, if you then exited all of `sharesIn` for tokenIn.Denom
+	estimateCoinOutGivenShares := func(sharesIn sdk.Int) (tokenOut sdk.Int, err error) {
```
Comment (p0mvn) [inline x/gamm/pool-models/internal/cfmm_common/lp.go:127]: What do you think about making this an actual function so that it can be unit tested?

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+		// new pool with added liquidity & LP shares, which we can mutate.
+		poolWithUpdatedLiquidity := poolWithAddedLiquidityAndShares(tokenIn, sharesIn)
+		swapToDenom := tokenIn.Denom
```
Comment (p0mvn) [inline x/gamm/pool-models/internal/cfmm_common/lp.go:130]:
```suggestion
```

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+		// so now due to correctness of exitPool, we exitPool and swap all remaining assets to base asset
+		exitFee := sdk.ZeroDec()
+		exitedCoins, err := poolWithUpdatedLiquidity.ExitPool(ctx, sharesIn, exitFee)
+		if err != nil {
+			return sdk.Int{}, err
+		}
+
+		return swapAllCoinsToSingleAsset(poolWithUpdatedLiquidity, ctx, exitedCoins, swapToDenom)
```
Comment (p0mvn) [inline x/gamm/pool-models/internal/cfmm_common/lp.go:138]:
```suggestion
		return swapAllCoinsToSingleAsset(poolWithUpdatedLiquidity, ctx, exitedCoins, tokenIn.Denom)
```

File: x/gamm/pool-models/stableswap/amm.go [hunk 11]
```diff
@@ -352,11 +356,51 @@ func (pa *Pool) calcInAmtGivenOut(tokenOut sdk.Coin, tokenInDenom string, swapFe
...
+// TODO: some day switch this to a COW wrapped pa, for better perf
+func (pa *Pool) joinPoolSharesInternal(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error) {
+	if len(tokensIn) == 1 {
+		numShares, err = pa.calcSingleAssetJoinShares(tokensIn[0], swapFee)
+		newLiquidity = tokensIn
+		return numShares, newLiquidity, err
+	} else if len(tokensIn) != pa.NumAssets() {
+		return sdk.ZeroInt(), sdk.NewCoins(), errors.New(
+			"stableswap pool only supports LP'ing with one asset, or all assets in pool")
+	}
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/amm.go:387]: nit: `else if` adds verbosity for input checking
```suggestion
	}
	if len(tokensIn) != pa.NumAssets() {
		return sdk.ZeroInt(), sdk.NewCoins(), errors.New(
			"stableswap pool only supports LP'ing with one asset, or all assets in pool")
	}
```

**Review by p0mvn — COMMENTED** [review]
Awesome work! 

I saw your question on Slack about splitting this. I think it would be great to get this approved and merged and continue the work in a separate PR / branch

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+	return numShares, remCoins, nil
+}
+
+// We binary search a number of LP shares, s.t. if we exited the pool with the updated liquidity,
+// and swapped all the tokens back to the input denom, we'd get the same amount. (under 0 swap fee)
+// Thanks to CFMM path-independence, we can estimate slippage with these swaps to be sure to get the right numbers here.
+// (by path-independence, swap all of B -> A, and then swap all of C -> A will yield same amount of A, regardless
+// of order and interleaving)
+//
+// This implementation all of pool.GetTotalPoolLiquidity, pool.ExitPool, and pool.SwapExactAmountIn to not
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:106]: The point is that we rely on calls to those pool functions _not_ modifying state, only the pool struct, despite the functions taking in the context.

(So it only supports a subset of pools, which is at the moment all of our current pools. I think our DeFi plans should remove this soon)

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+		}
+		coinShareRatios[i] = shareRatio
+	}
+
+	if minShareRatio.Equal(sdk.MaxSortableDec) {
+		return numShares, remCoins, errors.New("unexpected error in MaximalExactRatioJoin")
+	}
+
+	remCoins = sdk.Coins{}
+	numShares = minShareRatio.MulInt(totalShares).TruncateInt()
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:77]: yup exactly!

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+		}
+		exitedCoins = exitedCoins.Add(sdk.NewCoin(asset.Denom, exitAmt))
+	}
+
+	return exitedCoins, nil
+}
+
+// MaximalExactRatioJoin LP's the maximal amount of tokens in possible, and returns the number of shares that'd be
+// and how many coins would be left over.
+func MaximalExactRatioJoin(p types.PoolI, ctx sdk.Context, tokensIn sdk.Coins) (numShares sdk.Int, remCoins sdk.Coins, err error) {
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:53]: yeah, agreed. This is ideally moving around tests being worked on in Balancer, and then it applies for both our pools =)

File: x/gamm/pool-models/stableswap/amm.go [hunk 15]
```diff
@@ -352,11 +356,51 @@ func (pa *Pool) calcInAmtGivenOut(tokenOut sdk.Coin, tokenInDenom string, swapFe
...
+
+	// Add all exact coins we can (no swap). ctx arg doesn't matter for Stableswap
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/amm.go:389]: We should expect more of these functions going to be stateful! I think thats a key insight for AMMs we want to be enabling

File: osmoutils/binary_search_test.go [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBinarySearch(t *testing.T) {
+	// straight line function that returns input. Simplest to binary search on,
+	// binary search directly reveals one bit of the answer in each iteration with this function.
+	lineF := func(a sdk.Int) (sdk.Int, error) {
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply osmoutils/binary_search_test.go:13]: Yes there should be

File: x/gamm/pool-models/stableswap/amm.go (hunk 11 above)
Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/amm.go:387]: I don't follow? Why would you not prefer else if?

Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/amm.go:387]: My general view on this sort of code linting is either should have a readability reason, or a linter enforce it

**Comment by ValarDragon** [comment]
Thanks for the review on this!!

File: x/gamm/pool-models/stableswap/amm.go (hunk 11 above)
Comment (p0mvn) [reply x/gamm/pool-models/stableswap/amm.go:387]: My reasoning for this comment was the following:

In the case of if/else, we only execute A if condA is true
```
if condA {
   // Do A
} else if condB {
   // Do B
}
```
On the other hand,
```
if condA {
   // Do A
}

if condB {
   // Do B
}
```

It is possible to do both A and B independent of the value of `condA`.

Functionally, it does not make a difference in this example since both branches return. However, when reading the code, there is a small readability overhead for validating this subtle difference between `if/else` and `if`. Especially, since Go frequently uses the latter pattern for error handling.

I don't have a strong opinion on this so feel free to keep what you think is best :+1:

File: x/gamm/pool-models/stableswap/amm.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/gamm/pool-models/stableswap/amm.go:389]: Thanks for clarifying. Just trying to understand this more - what is an example of how we could utilize statefulness here? What kind of state could we store and for what features?

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+	if exitingShares.GTE(totalShares) {
+		return sdk.Coins{}, errors.New("too many shares out")
+	}
+
+	// refundedShares = exitingShares * (1 - exit fee)
+	// with 0 exit fee optimization
+	var refundedShares sdk.Dec
+	if !exitFee.IsZero() {
+		// exitingShares * (1 - exit fee)
+		// Todo: make a -1 constant
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:24]: idk why I had that comment, thanks! Deleted

File: x/gamm/pool-models/internal/cfmm_common/lp.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:118]: No, this choice needs to be analyzed. Not sure what you mean by extract up top? Like constants in the file?

File: x/gamm/pool-models/stableswap/amm.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/amm.go:389]: https://github.com/osmosis-labs/osmosis/issues/570

Also making volatility aware AMMs

File: x/gamm/pool-models/internal/cfmm_common/lp.go (hunk 6 above)
Comment (ValarDragon) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:118]: Needs to be revisited later

File: x/gamm/pool-models/internal/cfmm_common/lp.go (hunk 7 above)
Comment (ValarDragon) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:111]: We need functions that aren't on the pool API to do this.

File: x/gamm/pool-models/stableswap/amm.go [hunk 18]
```diff
@@ -352,11 +356,51 @@ func (pa *Pool) calcInAmtGivenOut(tokenOut sdk.Coin, tokenInDenom string, swapFe
...
+		paCopy := pa.Copy()
+		paCopy.updatePoolForJoin(sdk.NewCoins(tokenIn), newShares)
+		return &paCopy
+	}
+	// TODO: Correctly handle swap fee
+	return cfmm_common.BinarySearchSingleAssetJoin(pa, tokenIn, poolWithAddedLiquidityAndShares)
+}
+
+// We can mutate pa here
+// TODO: some day switch this to a COW wrapped pa, for better perf
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/amm.go:378]: Its a micro-optimization. I'm not sure it needs that much attention brought to it personally, I'd rather anyone interested in performance look elsewhere. Imo the comment mostly serves to indicate what we're weakly emulating with the internal function.

File: x/gamm/pool-models/internal/cfmm_common/lp.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:118]: Yes, suggesting to have these constants defined on top of the file:
```
const (
        binarySearchCorrectnessThreshold = 2
	binarySearchMaxIterations = 512
)
```

This helps to avoid defining these on every call to a function. Also, these constants can be reused in unit tests instead of redefined and potentially going out of sync with the actual value

File: x/gamm/pool-models/internal/cfmm_common/lp.go (hunk 7 above)
Comment (p0mvn) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:111]: So that we can customize them for each pool implementation if needed?

Comment (p0mvn) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:111]: If that's the reason, a potential alternative is to define a function on the pool interface and call it inside `BinarySearchSingleAssetJoin`

e.g.
```
func BinarySearchSingleAssetJoin(pool types.PoolI, ...) {
    ...
    poolWithUpdatedLiquidity := pool.GetCopyWithLiquidityAndShares(tokenIn, sharesIn)
    ...
}
```

File: x/gamm/pool-models/internal/cfmm_common/lp.go (hunk 8 above)
Comment (mattverse) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:127]: +1 on this comment!

File: x/gamm/pool-models/internal/cfmm_common/lp.go (hunk 4 above)
Comment (mattverse) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:140]: What's the exact task that this TODO is mentioning? Is it referring to creating test cases?

File: osmoutils/binary_search.go [hunk 19]
```diff
@@ -0,0 +1,95 @@
+package osmoutils
```
Comment (mattverse) [inline osmoutils/binary_search.go:1]: What's the standard for methods that go into `osmoutils` package and methods that stay within a specific package?

**Review by mattverse — APPROVED** [review]
Overall LGTM! 

I'm surprised how this PR kinda puts all the particles of stable swap together! 🚀

File: x/gamm/pool-models/internal/cfmm_common/lp.go (hunk 7 above)
Comment (ValarDragon) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:111]: Yes this is an alternative. I don't think it is preferrable

Comment (ValarDragon) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:111]: I think this adding such a method is premature abstraction.

File: osmoutils/binary_search.go [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+type ErrTolerance struct {
+	AdditiveTolerance       sdk.Int
+	MultiplicativeTolerance sdk.Dec
+}
+
+// IsSatisfied returns if actual is within errTolerance of expected.
+// returns 0 if it is
+// returns 1 if not, and expected > actual.
+// returns -1 if not, and expected < actual
+func (e ErrTolerance) Compare(expected sdk.Int, actual sdk.Int) int8 {
```
Comment (mconcat) [inline osmoutils/binary_search.go:24]: Some of the standard library `Compare` functions return `int` instead of `int8`. Functions taking compare function could expect type signature of return type `int`.
- https://pkg.go.dev/bytes#Compare
- https://pkg.go.dev/strings@go1.18.1#Compare
- https://pkg.go.dev/math/big@go1.18.1#Int.Cmp

File: osmoutils/binary_search.go [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// ErrTolerance is used to define a compare function, which checks if two
+// ints are within a certain error tolerance of one another.
+// ErrTolerance.Compare(a, b) returns true iff:
+// |a - b| <= AdditiveTolerance
+// |a - b| / min(a, b) <= MultiplicativeTolerance
+// Each check is respectively ignored if the entry is nil (sdk.Dec{}, sdk.Int{})
+type ErrTolerance struct {
```
Comment (mconcat) [inline osmoutils/binary_search.go:15]: Great to accompany both tolerance criteria!

**Review by mconcat — COMMENTED** [review]
Left some of comments!

File: osmoutils/binary_search.go (hunk 19 above)
Comment (ValarDragon) [reply osmoutils/binary_search.go:1]: In this case, they could plausibly belong in osmomath

File: x/gamm/pool-models/internal/cfmm_common/lp.go [hunk 22]
```diff
@@ -0,0 +1,163 @@
+package cfmm_common
```
[the remark this answers is not in the document]
Comment (mattverse) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:1]: import cycles cursed

File: x/gamm/pool-models/internal/cfmm_common/lp.go (hunk 1 above)
Comment (ValarDragon) [reply x/gamm/pool-models/internal/cfmm_common/lp.go:46]: exitAmt isn't a share amount. Its a `ratio * assets = assets`

File: x/gamm/pool-models/stableswap/amm.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/amm.go:389]: RIP if only we had further progress on volatility aware AMMs pre UST collapse :(
